atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/AbstractGetCandidateCellsForSweepingTest.java

```diff
@@ -179,6 +180,17 @@ public void startFromGivenRowThorough() {
                 .containsExactly(cell(2, 1), cell(2, 2), cell(3, 1), cell(3, 2));
     }
 
+    @Test
+    public void considersSentinelsForThorough() {
+        new TestDataBuilder()
+                .put(1, 1, -1L)
+                .put(1, 1, 5L)
+                .store();
+        CandidateCellForSweeping candidateCellForSweeping = Iterables.getOnlyElement(
```

Comment:
I don't think there's ever a case where a sentinel could be written but not 'committed' - is this correct? If that's the only case where this issue comes up, should we instead special case sentinels in the targeted sweep code to avoid the `TransactionService` lookups?

Reply:
This is true in standard Atlas usage. I guess the broader question is whether we want to break the behaviour of a `TransactionService`, which previously accepted negative numbers. This may be reasonable, since internal tests and large internal product tests haven't failed otherwise

Reply:
Writing of sentinels is nontransactionally done in sweep - it's done before deleting stuff, so there's no guarantee that we don't attempt to put sentinels multiple times, but it's guaranteed to happen at least once.

Reply:
So I think whether we want to continue to support or be robust to negative accesses on transactions service is still a bit of an open question. I'm not currently aware of any other places in the codebase where we access negative timestamps, and as mentioned the tests are favourable.

I've fixed up the sweeper to behave properly here, its behaviour on encountering sentinels in the past was pretty silly, anyway:

- it sees `-1` as a normal timestamp
- so it queries the db for `-1`, and finds that the "transaction" at `-1` either has been "rolled back" with `-1` or hasn't been
- if it hasn't, it tries to rollback the `-1` transaction by putting `-1, -1` into the transactions1 table
- then it reads that cell as corresponding to an uncommitted transaction

Now we filter out `-1` starts, and put them back later on (if it's thorough and we need to consider them)